pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BaseTableUpsertMetadataManager.java

```diff
@@ -75,6 +107,127 @@ public void init(TableConfig tableConfig, Schema schema, TableDataManager tableD
+      // Note that there is an implicit waiting logic between the thread doing the segment preloading here and the
+      // other helix threads about to process segment state transitions (e.g. taking segments from OFFLINE to ONLINE).
+      // The thread doing the segment preloading here must complete before the other helix threads start to handle
+      // segment state transitions. This is ensured implicitly because segment preloading happens here when
+      // initializing this TableUpsertMetadataManager, which happens when initializing the TableDataManager, which
+      // happens as the lambda of ConcurrentHashMap.computeIfAbsent() method, which ensures the waiting logic.
+      try {
+        _isPreloading = true;
+        preloadSegments();
+      } catch (Exception e) {
+        // Even if preloading fails, we should continue to complete the initialization, so that TableDataManager can be
+        // created. Once TableDataManager is created, no more segment preloading would happen, and the normal segment
+        // loading logic would be used. The segments not being preloaded successfully here would be loaded via the
+        // normal segment loading logic, the one doing more costly checks on the upsert metadata.
+        LOGGER.warn("Failed to preload segments for table: {}", _tableNameWithType, e);
+        if (e instanceof InterruptedException) {
+          // Restore the interrupted status in case the upper callers want to check.
+          Thread.currentThread().interrupt();
+        }
+      } finally {
+        _isPreloading = false;
+      }
+    }
+  }
+
+  /**
+   * Get the ideal state and find segments assigned to current instance, then preload those with validDocIds snapshot.
+   * Skip those without the snapshots and those whose crc has changed, as they will be handled by normal Helix state
+   * transitions, which will proceed after the preloading phase fully completes.
+   */
+  private void preloadSegments()
+      throws Exception {
+    LOGGER.info("Preload segments in table: {} for fast upsert metadata recovery", _tableNameWithType);
+    IdealState idealState = HelixHelper.getTableIdealState(_helixManager, _tableNameWithType);
+    ZkHelixPropertyStore<ZNRecord> propertyStore = _helixManager.getHelixPropertyStore();
+    String instanceId = getInstanceId();
+    IndexLoadingConfig indexLoadingConfig = createIndexLoadingConfig();
+    List<Future<?>> futures = new ArrayList<>();
+    for (String segmentName : idealState.getPartitionSet()) {
+      Map<String, String> instanceStateMap = idealState.getInstanceStateMap(segmentName);
+      String state = instanceStateMap.get(instanceId);
+      if (!CommonConstants.Helix.StateModel.SegmentStateModel.ONLINE.equals(state)) {
+        LOGGER.info("Skip segment: {} as its ideal state: {} is not ONLINE", segmentName, state);
+        continue;
+      }
+      if (_segmentPreloadExecutor == null) {
```

Comment:
(optional) Should we consider not allowing preloading when the executor is not configured? This will make it single threaded can potentially cause performance degradation